gr-uhd/lib/rfnoc_rx_streamer_impl.cc

```diff
@@ -0,0 +1,192 @@
+    }
+    return true;
+}
+
+bool rfnoc_rx_streamer_impl::stop()
+{
+    // If we issue a stream command on start, we also issue it on stop
+    if (d_issue_stream_cmd_on_start) {
+        ::uhd::stream_cmd_t stream_cmd(::uhd::stream_cmd_t::STREAM_MODE_STOP_CONTINUOUS);
+        d_streamer->issue_stream_cmd(stream_cmd);
+    }
+    flush();
+    return true;
+}
+
+int rfnoc_rx_streamer_impl::work(int noutput_items,
+                                 gr_vector_const_void_star& input_items,
+                                 gr_vector_void_star& output_items)
+{
+    const size_t max_num_items_to_rx = noutput_items * d_vlen;
+    const size_t num_items_recvd =
+        d_streamer->recv(output_items, max_num_items_to_rx, d_metadata, d_timeout);
+
+    const size_t num_vecs_recvd = num_items_recvd / d_vlen;
+    if (num_items_recvd % d_vlen) {
+        // TODO: Create a fix for this. What will happen is that a partial
+        // vector will be received, but it won't be available in the output_items.
+        // We need to store the partial vector, and prepend it to the next
+        // run.
+        GR_LOG_WARN(d_logger, "Received fractional vector! Expect signal fragmentation.");
+    }
+
+    using ::uhd::rx_metadata_t;
+    switch (d_metadata.error_code) {
+    case rx_metadata_t::ERROR_CODE_NONE:
+        break;
+
+    case rx_metadata_t::ERROR_CODE_TIMEOUT:
+        // its ok to timeout, perhaps the user is doing finite streaming
+        GR_LOG_DEBUG(d_logger, "UHD recv() call timed out.");
+        break;
+
+    case rx_metadata_t::ERROR_CODE_OVERFLOW:
+        // Not much we can do about overruns here, and they get signalled via the
+        // UHD logging interface
+        break;
+
+    default:
+        GR_LOG_WARN(
+            d_logger,
+            str(boost::format("RFNoC Streamer block received error %s (Code: 0x%x)") %
+                d_metadata.strerror() % d_metadata.error_code));
+    }
+
+    if (d_metadata.end_of_burst) {
+        for (size_t i = 0; i < output_items.size(); i++) {
+            add_item_tag(i, nitems_written(i) + num_vecs_recvd - 1, EOB_KEY, pmt::PMT_T);
+        }
+    }
+
+    return num_vecs_recvd;
+}
+
+/******************************************************************************
+ * rfnoc_rx_streamer API
+ *****************************************************************************/
+void rfnoc_rx_streamer_impl::set_start_time(const ::uhd::time_spec_t& time)
+{
+    d_start_time_set = true;
+    d_start_time = time;
+}
+
+/******************************************************************************
+ * Helpers
+ *****************************************************************************/
+void rfnoc_rx_streamer_impl::flush()
+{
+    constexpr size_t nbytes = 4096;
+    const size_t nchan = d_streamer->get_num_channels();
+    std::vector<std::vector<uint8_t>> buffs(nchan, std::vector<uint8_t>(nbytes));
+
+    gr_vector_void_star outputs;
```

Comment:
could have called the `std::vector` ctor that accepts a length, and avoided all the pushback's, but that's just a matter of preference in a low-frequency function like this :)